xspec/xspec
# Basic test for code coverage (Ant)

This pull request derives from #299. So needs to be handled after that.

---

This pull request adds some basic tests for the code coverage on Ant (implemented in #279).

Some of the tests are enabled only when Saxon (specified by `SAXON_JAR`) is capable of the code coverage. So they are skipped on Travis and AppVeyor. They are enabled automatically, if you're running the tests on your machine with Saxon license.

### How it works

* `run-bats.sh` (and `.cmd`) checks to see if `SAXON_JAR` can compile the coverage reporter XSL. 
If it can, the environment variable `XSLT_SUPPORTS_COVERAGE` is set. (d557f39fa0de45b2c329f35d108ef7404e1bb464)
* `xspec.bats` (and `collection.xml`) checks `XSLT_SUPPORTS_COVERAGE` before testing the coverage report feature. (01e638645c6535f53bf855d6b5413cf315a84154)

### How to test

Here's what I tested on my local machine.

1. Do `set SAXON_JAR=\path\to\saxon9ee.jar`. (`saxon9pe.jar` should be fine too, but I haven't tested.)
1. Run `run-bats.cmd` in `test/`.
    Must pass, including the test case 47.

1. Do `set SAXON_JAR=\path\to\saxon9he.jar`. (Notice the file name; It's Saxon-HE.)
1. Run `run-bats.cmd`.
    Must pass, skipping the test case 47. (demonstrated on [Travis](https://travis-ci.org/AirQuick/xspec/jobs/418981014#L1229) and [AppVeyor](https://ci.appveyor.com/project/AirQuick/xspec/build/757-test_ant-coverage/job/nv9tkyhouexj7l82#L335))

1. Check out 024c845c2a7b7e022987514458e74ffce08f3a36 which is intentionally broken.
1. Do `set SAXON_JAR=\path\to\saxon9ee.jar`.
1. Run `run-bats.cmd`.
    At least 29, 47, 48 and 49 must fail. (29, 48 and 49 are demonstrated on [Travis](https://travis-ci.org/AirQuick/xspec/builds/418985079) and [AppVeyor](https://ci.appveyor.com/project/AirQuick/xspec/build/758-test_ant-coverage))

Repeat the same steps using `run-bats.sh --tap` on Ubuntu.
